### ISSUE TITLE
references compared by reference/identity

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#10844]
### Context

Functions and lazy values cannot be compared or hash by the default, structural polymorphic functions. Using them inside some datatype forces users to write their own comparison/hashing function, which is painful. (This can be partly avoided by meta-programming / ppx_compare, but there remains a usability cliff when moving from polymorphic comparison to user- or meta-generated functions.)

One solution to this problem is to wrap the non-comparable value in a single-method object. Objects are not compared structurally, but "by identity" -- they store a unique integer identifier for this purpose. This works, but (1) it involves the object runtime when creating and dereferencing the wrapper, which has a non-trivial performance impact, and (2) it creates a 3-field block (class pointer, unique identifier, value), while a 2-field block would suffice (no need for a class pointer).

Remark: the code in #10825 would benefit from this feature, but cannot use objects for compiler-bootstrapping reasons. 

### Feature wish

@alainfrisch proposes to expose "references compared by identity" as a dedicated module with the following interface:

```ocaml
module PhysRef : sig
  type 'a t (* = private { mutable contents: 'a; uid: int } ?? *)
  val create: 'a -> 'a t
  val get: 'a t -> 'a
  val set: 'a t -> 'a -> unit
  (* val id: 'a t -> int ?? *)
end
```

Feature wish: expose such a PhysRef module in the standard library.

As a first step, we could implement it in the compiler codebase (not exposed to stdlib users) and see how we like it.

Remark: I personally have no use for the `set` operation and would be content with immutable references, but I assume that Lexifi also has a use for mutation.

### Implementation

In https://github.com/ocaml/ocaml/pull/10825#issuecomment-1004137463, @garrigue and @alainfrish offer two potential solutions to this problem, which can be implemented in user code without runtime change:
- @garrigue suggests an object-like layout with a unit value instead of the class pointer
- @alainfrisch mentions a variant used in production by Lexifi, which is a 2-field block with a well-chosen tag
 
(@alainfrisch mentions tag 252, which iirc. corresponds to String_tag, which sounds like dark magic, I would have assumed that they use Object_tag (248) instead.)

I wondered if either proposal could create trouble for the runtime, which may assume things about the layout of Object_tag value. But the runtime already uses the two-field Object_tag layout to represent exceptions and extensible variants, so it cannot make too many assumptions about Object_tag blocks; I believe that either proposal (using Object_tag) would be safe.

In other words, the proposal would generalize the meaning of `Object_tag` values to be a specialized form of custom blocks with arbitrary-many (>1) user-chosen fields plus a unique identifier. Extensible variant tags correspond to values of type `string phys_ref`.

Note: the Multicore runtime needs a global atomic counter to generate unique identifiers, but this should be fast enough -- we already need this global atomic counter with objects and extensible variants anyway.

**Comment by alainfrisch** [comment]
> (@alainfrisch mentions tag 252, which iirc. corresponds to String_tag, which sounds like dark magic, I would have assumed that they use Object_tag (248) instead.)

That was obviously a typo, I meant Object_tag.

> Remark: I personally have no use for the set operation and would be content with immutable references, but I assume that Lexifi also has a use for mutation.

To be clear, no, we don't use such mutable references.  We use the implementation technique in several places, keeping more than one field of data in the records.  To support this use case without the current hacks (involving `Obj.set_tag/with_tag`), one could use an attribute on record declaration, to instruct the compiler to directly use Object_tag instead of 0 for those records (after checking that the second field is an immediate value).  I did not suggest that because it feels a bit too ugly.

**Comment by gasche** [comment]
I think that `Obj.with_tag` is in fact fine / efficient enough for most purposes (it's a copy of a short block), we probably don't need to implement additional compiler support unless there is a demonstrated need.

**Comment by alainfrisch** [comment]
I was not concerned by performance, but by the fact that users need to use such low-level, undocumented hack that relies on an unsafe feature.  But the compiler support is probably way too ad hoc anyway...

**Comment by xavierleroy** [comment]
> I wondered if either proposal could create trouble for the runtime, which may assume things about the layout of Object_tag value. 

The only assumption made about blocks carrying `Object_tag` is that they have size >= 2 and that the second field the object identifier (a unique tagged integer).

> Note: the Multicore runtime needs a global atomic counter to generate unique identifiers, but this should be fast enough 

Actually, Multicore has per-domain counters that are refreshed every 1024 steps (IIRC) from a global counter.  So it's about as efficient as you can get.

**Comment by gasche** [comment]
For the record, the following implementation seems to work as intended.

```ocaml
module PhysRef : sig
  type 'a t
  val create: 'a -> 'a t
  val get: 'a t -> 'a
  val id: 'a t -> int
end = struct
  type 'a t

  external fresh_id : unit -> int = "caml_fresh_oo_id" [@@noalloc]

  let create (type a) (v : a) : a t =
    Obj.obj (Obj.with_tag Obj.object_tag
      (Obj.repr (v, fresh_id ())))
      
  let get (type a) (r : a t) : a =
    Obj.obj (Obj.field (Obj.repr r) 0)

  let id (type a) (r : a t) : int =
    Obj.obj (Obj.field (Obj.repr r) 1)
end
```

(One downside of this approach is that `PhysRef.t` values are shown as `<abstr>` in the toplevel.)

This would be useful as a separate third-party library, but I don't see it moving to the stdlib very soon, so I'm closing the issue -- the most important aspect is that it is easy to find the code again if we need it.
